File: repository-blueprints/memebank/mb-clients/clients/typescript/src/runtime.ts
import type { ClientOperation, RetryClass } from "./generated.js";

export const defaultTimeoutMs = 30_000;
export const maxAttempts = 3;
export const redacted = "[REDACTED]";

export interface AccessTokenProvider {
  accessToken(signal?: AbortSignal): Promise<string>;
  refreshAccessToken(signal?: AbortSignal): Promise<string>;
}

export interface RequestPlan {
  readonly operation: ClientOperation;
  readonly path: string;
  readonly headers: Readonly<Record<string, string>>;
  readonly body?: ReadableStream<Uint8Array>;
  readonly timeoutMs: number;
  readonly signal?: AbortSignal;
}

export interface TransportResponse {
  readonly status: number;
  readonly bodyStarted: boolean;
}

export interface ClientTransport {
  execute(plan: RequestPlan): Promise<TransportResponse>;
}

export function shouldRetry(input: {
  retryClass: RetryClass;
  status: number;
  attempt: number;
  bodyStarted: boolean;
}): boolean {
  if (input.retryClass === "never" || input.attempt >= maxAttempts || input.bodyStarted) {
    return false;
  }
  return new Set([408, 425, 429, 500, 502, 503, 504]).has(input.status);
}

export function redactHeaders(
  headers: Readonly<Record<string, string>>,
): Readonly<Record<string, string>> {
  const sensitive = new Set([
    "authorization",
    "cookie",
    "set-cookie",
    "x-api-key",
    "x-provider-token",
  ]);
  return Object.fromEntries(
    Object.entries(headers)
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([name, value]) => [name, sensitive.has(name.toLowerCase()) ? redacted : value]),
  );
}

File: repository-blueprints/memebank/mb-clients/clients/dart/lib/src/runtime.dart
import 'generated.dart';

const defaultTimeout = Duration(seconds: 30);
const maxAttempts = 3;
const redacted = '[REDACTED]';

abstract interface class AccessTokenProvider {
  Future<String> accessToken();
  Future<String> refreshAccessToken();
}

abstract interface class ClientTransport {
  Future<TransportResponse> execute(RequestPlan request);
}

final class TransportResponse {
  const TransportResponse({required this.statusCode, required this.bodyStarted});

  final int statusCode;
  final bool bodyStarted;
}

final class RequestPlan {
  const RequestPlan({
    required this.operation,
    required this.path,
    required this.headers,
    this.body,
    this.timeout = defaultTimeout,
  });

  final ClientOperation operation;
  final String path;
  final Map<String, String> headers;
  final Stream<List<int>>? body;
  final Duration timeout;
}

bool shouldRetry({
  required RetryClass retryClass,
  required int status,
  required int attempt,
  required bool bodyStarted,
}) {
  if (retryClass == RetryClass.never || attempt >= maxAttempts || bodyStarted) {
    return false;
  }
  return const {408, 425, 429, 500, 502, 503, 504}.contains(status);
}

Map<String, String> redactHeaders(Map<String, String> headers) {
  const names = {
    'authorization',
    'cookie',
    'set-cookie',
    'x-api-key',
    'x-provider-token',
  };
  return {
    for (final entry in headers.entries)
      entry.key: names.contains(entry.key.toLowerCase()) ? redacted : entry.value,
  };
}
